### PR TITLE
Derive max composite buffers from max content len

### PR DESCRIPTION
**Pull request by danielmitterdorfer**
With this commit we determine the maximum number of buffers that Netty
keeps while accumulating one HTTP request based on the maximum content
length. Previously, we kept the default value of 1024 which is too small
for bulk requests which leads to unnecessary copies of byte buffers
internally.

### CODE REVIEW

**Comment by danielmitterdorfer** [comment]
I have now run several benchmarks in our benchmarking suite with different heap sizes. Below are the results for two extreme cases:

* `geopoint` which contains only very small documents. I ran the benchmarks with an increased bulk size of 50.000 docs per bulk (instead of 5.000).
* `pmc` where each document is one academic paper. `pmc` is rather I/O intensive and has large bulk requests compared to `geopoint`.

I ran these benchmarks with different heap sizes to see for which configurations the max cumulation buffer changes have an effect. The results match the expectation that with these change, Elasticsearch performs better under low memory conditions while performing similar to or slightly better than before when more heap memory is available. One data point where this is very apparent is that we could successfully finish `pmc` with 768MB heap with these changes whereas the benchmark candidate OOMEd after indexing 30% of the document corpus without these changes (indicated with indexing thoughput set to zero in the graph below).

The graphs below show the achieved median indexing throughput for those tracks along with error bars for the minimum and maximum throughput. In blue we see the current default value of at most 1024 cumulation buffer components whereas in orange we see the new default of 69905 cumulation buffer components assuming a default HTTP max content length of 100MB (this value is derived based on the HTTP max content length; see the code for details).

throughput_geopoint

throughput_pmc

I'd intend to merge these changes to master after successful review, let it bake there for a while and backport it to 6.x.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java [hunk 1]
```diff
@@ -122,8 +122,34 @@
...
+            // With this setting we control the CompositeByteBuf's capacity (which is by default 1024, see
+            // io.netty.handler.codec.MessageAggregator#DEFAULT_MAX_COMPOSITEBUFFER_COMPONENTS). To determine a proper default capacity for
+            // that buffer, we need to consider that the upper bound for the size of HTTP requests is determined by `maxContentLength`. The
+            // number of buffers that are needed depend on how often Netty reads network packets which depends on the network type (MTU).
+            // We assume here that Elasticsearch receives HTTP requests via an Ethernet connection which has a MTU of 1500 bytes.
+            //
+            // Note that we are *not* pre-allocating any memory based on this setting but rather determine the CompositeByteBuf's capacity.
+            // The tradeoff is between less (but larger) buffers that are contained in the CompositeByteBuf and more (but smaller) buffers.
+            // With the default max content length of 100MB and a MTU of 1500 bytes we would allow 69905 entries.
+            long maxBufferComponentsEstimate = Math.round((double) (maxContentLength.getBytes() / MTU_ETHERNET.getBytes()));
```
Comment (jasontedor) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java]: We can get the interface MTU from `NetworkInterface` API; should we use this here?

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java [hunk 2]
```diff
@@ -122,8 +122,34 @@
...
+            //
+            // With this setting we control the CompositeByteBuf's capacity (which is by default 1024, see
+            // io.netty.handler.codec.MessageAggregator#DEFAULT_MAX_COMPOSITEBUFFER_COMPONENTS). To determine a proper default capacity for
+            // that buffer, we need to consider that the upper bound for the size of HTTP requests is determined by `maxContentLength`. The
+            // number of buffers that are needed depend on how often Netty reads network packets which depends on the network type (MTU).
+            // We assume here that Elasticsearch receives HTTP requests via an Ethernet connection which has a MTU of 1500 bytes.
+            //
+            // Note that we are *not* pre-allocating any memory based on this setting but rather determine the CompositeByteBuf's capacity.
+            // The tradeoff is between less (but larger) buffers that are contained in the CompositeByteBuf and more (but smaller) buffers.
+            // With the default max content length of 100MB and a MTU of 1500 bytes we would allow 69905 entries.
```
Comment (jasontedor) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java]: It is a nit, but would you please use the multi-line style:

```
/*
 *
 */
```?

**Review by jasontedor — COMMENTED** [review]
I left a question.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java (hunk 2 above)
Comment (danielmitterdorfer) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java]: Sure. Addressed in https://github.com/elastic/elasticsearch/pull/29448/commits/b5c14fd7af00803293c47d69877c8991a083c789.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java (hunk 1 above)
Comment (danielmitterdorfer) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java]: I had a look this and I really like your idea. However, I am not sure it is feasible: I guess we need to resolve the network interface that is associated with the publish address (leveraging `NetworkService`). 

While I also dislike having the MTU specified as a constant here (instead of retrieving it from the network interface), it serves as an input for an estimation. The "worst" practical MTU is 1500 byte, another typical one is 65536 byte for loopback (that's not always the case for loopback but it is usually higher than for Ethernet). What happens if we assume an MTU of 1500 for our estimation in the loopback case? Our estimate would allow more buffer components than are expected for loopback. But in practice we do not expect to reach that theoretical capacity. A specific example for the loopback case (with a max content length of 100 MB):

* Estimated maximum number of buffers: 100 MB / 1500 byte = 69905
* Actual maximum number of buffers: 100 MB / 65536 byte = 1600

This means that we expect to reserve space for at most 1600 buffers in the `HttpObjectAggregator` but overestimate it to at most 69905 (which we do not reach in practice).

To summarize, I see two possibilities:

* Determine the publish address with `NetworkService`, find the matching network interface and determine its MTU. This is correct but more complex.
* Rename the constant `MTU_ETHERNET` to something like `SMALLEST_EXPECTED_MTU` (while [RFC 791](https://tools.ietf.org/html/rfc791) states 68 bytes as the minimum, I'd stick to 1500 for practical purposes). It's not ideal but in very special cases, users can explicitly define this expert setting (`http.netty.max_composite_buffer_components`) directly.

Comment (danielmitterdorfer) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java]: @jasontedor wdyt about the above?

Comment (jasontedor) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java]: I think that jumbo frames are common enough that we should try to go the extra mile here. If it's not possible to do cleanly, I would at least like to see a system property that can be set to set the MTU to 9000.

**Comment by tbrooks8** [comment]
At some point it may be worth benchmarking this with TLS/SSL enabled as some of the assumptions (1500 byte buffer sizes making it to `HttpObjectAggregator`) may not hold. A full TLS/SSL packet (which could be 16kb) will need to be decrypted prior to doing http decoding. The risk of course is that larger component lists consume more heap. But I think it will not be a problem as Netty seems to start with lists of 16 components in the `CompositeByteBufs` and resizes as necessary. Still, might be worth checking. Or could just monitor security benchmarks after merge.

**Comment by danielmitterdorfer** [comment]
Good point @tbrooks8. I ran the PMC benchmarks x-pack now and the results are similar to Elasticsearch without x-pack (i.e. we still see a benefit of implementing this change):

throughput_pmc x-pack

> The risk of course is that larger component lists consume more heap.

The number that is adjusted here defines at which point the original (small) buffers get copied into one larger buffer. We do not preallocate any memory based on that number. I agree that we might have more elements in the `CompositeByteBuf` but based on the benchmark results I'd conclude that more entries the internal array list are less problematic than the additional buffer copies.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java (hunk 1 above)
Comment (danielmitterdorfer) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java]: Thanks for the suggestion. I have now introduced a system property with a default is 1500 (bytes).

**Comment by danielmitterdorfer** [comment]
@elasticmachine test this.

**Comment by danielmitterdorfer** [comment]
@elasticmachine retest this please

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java [hunk 3]
```diff
@@ -122,8 +122,43 @@
...
     }
 
+    /*
+     * Size in bytes of an individual message received by io.netty.handler.codec.MessageAggregator which accumulates the content for an
+     * HTTP request. This number is used for estimating the maximum number of allowed buffers before the MessageAggregator's internal
+     * collection of buffers is resized.
+     *
+     * By default we assume the Ethernet MTU (1500 bytes) but users can override it with a system property.
+     */
+    private static final ByteSizeValue MTU = new ByteSizeValue(Long.valueOf(System.getProperty("es.net.mtu", "1500")));
```
Comment (jasontedor) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java]: Nit: it does not really matter here but this boxes and this should instead be `Long.parseLong`.

**Review by jasontedor — APPROVED** [review]
I left one comment but LGTM to me now.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java (hunk 3 above)
Comment (danielmitterdorfer) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java]: Thanks, I fixed it in 635c9fb.

**Comment by danielmitterdorfer** [comment]
Thanks for review @jasontedor and your comments @tbrooks8! I let this bake on master for a while and will then backport to 6.x.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java (hunk 3 above)
Comment (jasontedor) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java]: 🚢

**Comment by danielmitterdorfer** [comment]
@elasticmachine retest this please

**Comment by jasontedor** [comment]
@danielmitterdorfer I do not think this change should be backported to the 6.3 branch.

**Comment by danielmitterdorfer** [comment]
Agreed w.r.t releases. Backported to 6.x in 71d3297658e89aecd407842419ab21363989b1e3.
